File: src/main/java/nsu/MobileNetOptimiztion/NetOptimizationServer/Controller/CoverageNet.java
package nsu.MobileNetOptimiztion.NetOptimizationServer.Controller;

import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.DaoManager;
import com.j256.ormlite.jdbc.JdbcConnectionSource;
import nsu.MobileNetOptimiztion.NetOptimizationServer.DB.ConnectionStatistic;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import nsu.MobileNetOptimiztion.NetOptimizationServer.RequestCoverageNet;

import java.sql.SQLException;
import java.util.LinkedList;
import java.util.List;



public class CoverageNet {
    static final double KILOMETERS_IN_ONE_DEGREE_MERIDIAN = 111.134861111/2;
    static final double KILOMETERS_IN_ONE_DEGREE_ECVATOR = 111.321377778/2;


    private Dao<ConnectionStatistic, Long> dao;
    JdbcConnectionSource connectionSource;
    private float longtitude;
    private float latitude;
    private String connectionType;
    private int netSizeM = 5000;
    private int netStepM = 50;

    public CoverageNet(RequestCoverageNet requestCoverageNet) throws SQLException {
        longtitude = Float.parseFloat(requestCoverageNet.getLongtitude());
        latitude = Float.parseFloat(requestCoverageNet.getLat());
        connectionType = requestCoverageNet.getConnectionType();
        connectionSource = new JdbcConnectionSource("jdbc:postgresql://localhost/netoptimization","postgres","1901Beta");
        dao = DaoManager.createDao(connectionSource, ConnectionStatistic.class);
    }

    public String createCoverageNet() throws JSONException, SQLException {
        JSONObject requestBody = new JSONObject();
        int[] data = new int[0];
        String type = "";
        final float dLat = findDeltaLat(netStepM);
        final float dLong = findDeltaLong(netStepM);
        final float startLat = latitude - findDeltaLat(netSizeM);
        final float startLong = longtitude - findDeltaLong(netSizeM);
        final float finalLat = latitude + findDeltaLat(netSizeM);
        final float finalLong = longtitude + findDeltaLong(netSizeM);
        final int arraySize = netSizeM/netStepM;
        requestBody.put("dlat",dLat);
        requestBody.put("dlong",dLong);
        requestBody.put("startlat",startLat);
        requestBody.put("endlat",finalLat);
        requestBody.put("startlong",startLong);
        requestBody.put("endlong",finalLong);
        List<ConnectionStatistic> dataList = dao.queryBuilder()
                .where().eq("connection_type", connectionType)
                .query();
        LinkedList<String> sigStrengthList = new LinkedList<String>();
        for(ConnectionStatistic cur:dataList){
            sigStrengthList.add(cur.getSig_strength());
        }

        switch (connectionType){
            case "GSM":
                type = "rssi";
                data = makeGSMNet(dataList);
                break;
            case "CDMA": break;
            case "LTE":break;
            case "WCDMA":break;
            default:break;
        }
        requestBody.put("index_type",type);
        requestBody.put("size",arraySize);
        JSONArray array = new JSONArray(data);
        requestBody.put("data", array);
        return requestBody.toString();
    }
    private int[] makeGSMNet(List<ConnectionStatistic> dataList) {
        final float dLat = findDeltaLat(netStepM);
        final float dLong = findDeltaLong(netStepM);
        final float startLat = latitude - findDeltaLat(netSizeM);
        final float startLong = longtitude - findDeltaLong(netSizeM);
        final float finalLat = latitude + findDeltaLat(netSizeM);
        final float finalLong = longtitude + findDeltaLong(netSizeM);
        System.out.println("DLAT: " + dLat);
        System.out.println("DLong: " + dLong);
        System.out.println("Start lat: " + startLat);
        System.out.println("Final lat: " + finalLat);
        System.out.println("Start long: " + startLong);
        System.out.println("Final long: " + finalLong);
        final int arraySize = netSizeM/netStepM;
        int counters[] = new int[arraySize*arraySize];
        int data[] = new int[arraySize*arraySize];
        for(ConnectionStatistic stat:dataList){
            final float curLat = Float.parseFloat(stat.getLat());
            final float curLong = Float.parseFloat(stat.getLongtitude());
            final int rssi = getGSMRSSI(stat);
            if(curLat > finalLat || curLat < startLat) {
                System.out.println("Cur lat: " + curLat);
                continue;
            }
            if(curLong > finalLong || curLong < startLong) {
                System.out.println("Cur long: " + curLong);
                continue;
            }
            final int latId = (int) (0.6*(curLat - startLat)/dLat);
            final int longId = (int)(0.6*(curLong - startLong)/dLong);
            if(latId >= arraySize || longId >= arraySize || latId < 0 || longId < 0) {
                System.out.println("LatID: " + latId);
                System.out.println("LongID: " + longId);
                continue;
            }
//            System.out.println("LatID: " + latId);
//            System.out.println("LongID: " + longId);
//            System.out.println("RSSI: " + rssi);
            int offset = latId * arraySize + longId;
            counters[offset]++;
            data[offset] = (data[offset] + rssi)/counters[offset];
            if((longId + 1) < arraySize){
                offset = latId * arraySize + longId + 1;
                counters[offset]++;
                data[offset] = (data[offset] + rssi)/counters[offset];
            }
            if((latId + 1) < arraySize) {
                offset = (latId + 1) * arraySize + longId;
                counters[offset]++;
                data[offset] = (data[offset] + rssi)/counters[offset];
            }
            if((latId + 1) < arraySize && (longId + 1) < arraySize){
                offset = (latId + 1) * arraySize + longId + 1;
                counters[offset]++;
                data[offset] = (data[offset] + rssi)/counters[offset];
            }
        }
//        for(int i = 0; i < arraySize; i++){
//            for(int j = 0; j < arraySize; j++){
//                System.out.print(" " + data[i*arraySize + j] + " ");
//            }
//            System.out.println();
//            System.out.println("____________________");
//        }
        return data;
    }

    private int getGSMRSSI(ConnectionStatistic statistic){
        if(!statistic.getSig_strength().contains("rssi="))
            return 0;
        String rssiString = statistic.getSig_strength().substring(statistic.getSig_strength().indexOf("rssi="));
        rssiString = rssiString.substring(rssiString.indexOf("=")+1,rssiString.indexOf(" "));
        return Integer.parseInt(rssiString);
    }

    private float findDeltaLat(int distInMeters){
        double degreesInKilometer = 0.6/KILOMETERS_IN_ONE_DEGREE_MERIDIAN;
        float result =  (float) (distInMeters*degreesInKilometer/1000);
        //System.out.println("Result dLat: " + result);
        return result;
    }

    private float findDeltaLong(int distInMeters){
        double degreesInKilometer = 0.6/(Math.cos(latitude *Math.PI/180)*KILOMETERS_IN_ONE_DEGREE_ECVATOR);
        float result = (float)(distInMeters*degreesInKilometer/1000);
        //System.out.println("Result dLong: " + result);
        return result;
    }
}

File: src/main/java/nsu/MobileNetOptimiztion/NetOptimizationServer/Controller/CoverageNet.kt
package nsu.MobileNetOptimiztion.NetOptimizationServer.Controller

import com.google.gson.Gson
import com.j256.ormlite.dao.Dao
import com.j256.ormlite.dao.DaoManager
import com.j256.ormlite.jdbc.JdbcConnectionSource
import nsu.MobileNetOptimiztion.NetOptimizationServer.DB.ConnectionStatistic
import nsu.MobileNetOptimiztion.NetOptimizationServer.RequestEntities.CoverageNetRequest
import nsu.MobileNetOptimiztion.NetOptimizationServer.RequestEntities.CoverageRequestEntity
import nsu.MobileNetOptimiztion.NetOptimizationServer.Utils.GeoUtils
import nsu.MobileNetOptimiztion.NetOptimizationServer.Utils.GeoVec
import nsu.MobileNetOptimiztion.NetOptimizationServer.Utils.Location
import org.json.JSONException
import java.sql.SQLException
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class CoverageNet(private val netProperties: CoverageRequestEntity) {
    private val dao: Dao<ConnectionStatistic, Long>
    var connectionSource: JdbcConnectionSource

    @Throws(JSONException::class, SQLException::class)
    fun createCoverageNet(): String {
        var data = IntArray(0)
        var type = ""
        val dataList = dao.queryBuilder()
            .where().eq("connection_type", netProperties.connectionType)
            .query()
        return Gson().toJson(makeNet(dataList, getSSGetter(netProperties.connectionType)))
    }

    private fun <T> create2DArray(sizeX: Int, sizeY: Int): ArrayList<ArrayList<T>> {
        val arrayOfArray = ArrayList<ArrayList<T>>(sizeY)
        for (i in 0..sizeY - 1) {
            arrayOfArray[i] = ArrayList(sizeX)
        }
        return arrayOfArray
    }

    private fun interpolateColumn(net: ArrayList<ArrayList<Location>>, index: Int) {
        val lastIndex = net.size - 1
        val first = net[0][index]
        val last = net[lastIndex][index]
        val dlat = (first.latitude - last.latitude) / net.size
        val dlon = (first.latitude - last.latitude) / net.size
        for (i in 0..lastIndex) {

        }
    }

    private fun interpolateNet(
        net: ArrayList<ArrayList<Location>>,
        p00: Location,
        p01: Location,
        p10: Location,
        p11: Location,
    ) {
        val sizeY = net.size
        val sizeX = net[0].size
        net[0][0] = p00
        net[sizeY - 1][0] = p01
        net[0][sizeX - 1] = p10
        net[sizeY - 1][sizeX - 1] = p11

    }

    private fun getSSGetter(connectionType: String):(statistic: ConnectionStatistic)->Int =
        when (connectionType){
            "GSM" -> {a -> getGSMRSSI(a) }
            //"CDMA" -> { }
            //"LTE" -> { }
            "WCDMA" -> {a -> getWCDMASS(a) }
            else -> {a -> 0}
        }

    private fun makeNet(dataList: List<ConnectionStatistic>, ssGetter:(statistic: ConnectionStatistic)->Int): CoverageNetRequest {
        val loc00 = Location(netProperties.lat00.toFloat(), netProperties.lon00.toFloat())
        val loc11 = Location(netProperties.lat11.toFloat(), netProperties.lon11.toFloat())
        val geoVec00 = GeoVec(loc00)
        val geoVec01 = GeoVec(Location(loc11.latitude, loc00.longtitude))
        val geoVec10 = GeoVec(Location(loc00.latitude, loc11.longtitude))

        val sizeX = abs((GeoUtils.distanceCartesian(geoVec00, geoVec10) / netProperties.step).toInt())
        val sizeY = abs((GeoUtils.distanceCartesian(geoVec00, geoVec01) / netProperties.step).toInt())
        println("${sizeX} ${sizeY}")
        val dLat = (loc11.latitude - loc00.latitude) / sizeY
        val dLong = (loc11.longtitude - loc00.longtitude) / sizeX

        val counters = IntArray(sizeX * sizeY) { 0 }
        val data = IntArray(sizeX * sizeY) { 0 }
        dataList.filter {
            it.lat.toFloat() > loc00.latitude && it.lat.toFloat() < loc11.latitude &&
                    it.longtitude.toFloat() > loc00.longtitude && it.longtitude.toFloat() < loc11.longtitude
        }.forEach { stat ->
            val curLat = stat.lat.toFloat()
            val curLong = stat.longtitude.toFloat()
            val ss = ssGetter(stat)
            val latId = ((curLat - loc00.latitude) / dLat).toInt()
            val longId = ((curLong - loc00.longtitude) / dLong).toInt()
            val offset = latId * sizeX + longId
            counters[offset]++
            data[offset] = (data[offset] + ss)
        }
        for (i in 0 until sizeX * sizeY) {
            if (counters[i] > 0)
                data[i] = data[i] / counters[i]
        }
        return CoverageNetRequest(sizeX, sizeY, data, "rssi")
    }

    private fun getGSMRSSI(statistic: ConnectionStatistic): Int {
        if (!statistic.sig_strength.contains("rssi=")) return 0
        var rssiString = statistic.sig_strength.substring(statistic.sig_strength.indexOf("rssi="))
        rssiString = rssiString.substring(rssiString.indexOf("=") + 1, rssiString.indexOf(" "))
        return rssiString.toInt()
    }

    private fun getWCDMASS(statistic: ConnectionStatistic): Int {
        if (!statistic.sig_strength.contains("ss=")) return 0
        var ssString = statistic.sig_strength.substring(statistic.sig_strength.indexOf("ss="))
        ssString = ssString.substring(ssString.indexOf("=") + 1, ssString.indexOf(" "))
        return ssString.toInt()
    }

    init {
        connectionSource = JdbcConnectionSource("jdbc:postgresql://localhost/netoptimization", "postgres", "1901Beta")
        dao = DaoManager.createDao(
            connectionSource,
            ConnectionStatistic::class.java
        )
    }
}